Setter for real and imag tensor attributes

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39860 Setter for real and imag tensor attributes**

Differential Revision: [D22163234](https://our.internmc.facebook.com/intern/diff/D22163234)

anjali411 committed:
Setter for real and imag tensor attributes

[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

mruberry commented on test/test_torch.py:
```diff
@@ -18073,6 +18073,30 @@ def compare_with_numpy(contiguous_input=True):
+    def test_set_real_imag(self, device, dtype):
+        x = torch.randn(10, dtype=dtype)
+
+        for d in torch.testing.get_all_dtypes():
```
You can use

```
@dtypes(*product(torch.testing.get_all_complex_dtypes(), torch.testing.get_all_dtypes()))
```

and the third argument in your signature will be `dtypes` and it will be an element from the product (a tuple of two elements where dtypes[0] is from torch.testing.get_all_complex_dtypes() and dtypes[1] is from torch.testing.get_all_dtypes()).

mruberry commented on test/test_torch.py:
```diff
@@ -18073,6 +18073,30 @@ def compare_with_numpy(contiguous_input=True):
+                    x.real = new_real
+                with self.assertWarnsRegex(UserWarning, regex):
+                    x.imag = new_imag
+                self.assertEqualIgnoreType(x.real, new_real.real)
```
self.assertEqual(..., exact_dtype=False)

mruberry commented on test/test_torch.py:
```diff
@@ -18073,6 +18073,30 @@ def compare_with_numpy(contiguous_input=True):
+                x.real = new_real
+                x.imag = new_imag
+                self.assertEqualIgnoreType(x.real, new_real)
+                self.assertEqualIgnoreType(x.imag, new_imag)
```
Same here. assertEqualIgnoreType will eventually be removed, so we don't want to increase uses of it.

mruberry commented on test/test_torch.py:
```diff
@@ -18073,6 +18073,30 @@ def compare_with_numpy(contiguous_input=True):
+    @onlyOnCPUAndCUDA
+    @dtypes(*torch.testing.get_all_complex_dtypes())
+    def test_set_real_imag(self, device, dtype):
+        x = torch.randn(10, dtype=dtype)
```
`dtype=detype, device=device`

mruberry commented on test/test_torch.py:
```diff
@@ -18395,7 +18419,7 @@ def _make_tensor(shape, dtype, device, fill_ones=False) -> torch.Tensor:
 
     # Returns a tensor with random integer values
-    if not dtype.is_floating_point:
+    if not dtype.is_floating_point or not dtype.is_complex:
```
`if !(dtype.is_floating_point and dtype.is_complex)`

Won't your current check always go down this codepath?

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

mruberry commented:
Additional question: how will this work in C++? What's up with the derivatives.yaml entries for real and imag being removed by the view_as_real and view_as_complex PR? Will you be able to backward after performing .real?

albanD commented:
I don't know about the c++ api.
But "Will you be able to backward after performing .real?" the answer is yes. `at::real()` is now a composite function. And so only the derivatives of its constituents is needed (view_as_real + select).

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"




[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"


Differential Revision: [D22163234](https://our.internmc.facebook.com/intern/diff/D22163234)

[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"


Differential Revision: [D22163234](https://our.internmc.facebook.com/intern/diff/D22163234)

[ghstack-poisoned]

mruberry reviewed (approved):
LGTM!

anjali411 committed:
Update on "Setter for real and imag tensor attributes"


Differential Revision: [D22163234](https://our.internmc.facebook.com/intern/diff/D22163234)

[ghstack-poisoned]

anjali411 committed:
Update on "Setter for real and imag tensor attributes"


Differential Revision: [D22163234](https://our.internmc.facebook.com/intern/diff/D22163234)

[ghstack-poisoned]
